File: src/main/java/net/moltendorf/Bukkit/IntelliDoors/controller/AbstractDoor.java
package net.moltendorf.Bukkit.IntelliDoors.controller;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.Sound;

/**
 * Created by moltendorf on 16/5/2.
 */
public abstract class AbstractDoor implements Door {
  protected abstract Location getLocation();

  @Override
  public void wasToggled(Door onDoor) {
    if (getType() == Material.IRON_DOOR_BLOCK) {
      Location location = getLocation();

      if (isClosed()) {
        open();

        location.getWorld().playSound(location, Sound.BLOCK_IRON_DOOR_OPEN, 1, 1);
      } else {
        close();

        location.getWorld().playSound(location, Sound.BLOCK_IRON_DOOR_CLOSE, 1, 1);
      }
    }
  }
}

File: src/main/kotlin/net/moltendorf/Bukkit/IntelliDoors/controller/AbstractDoor.kt
package net.moltendorf.Bukkit.IntelliDoors.controller

import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.Block

/**
 * Created by moltendorf on 16/5/2.
 */
abstract class AbstractDoor(val block: Block) : Door() {
  override val location: Location
    get() {
      return block.location
    }

  override val type: Material
    get() {
      return block.type
    }

  protected var data: Int = block.data.toInt()

  override var open: Boolean
    get() {
      return data and 4 == 0
    }
    set(value) {
      if (open != value) {
        data += if (value) 4 else -4
        block.data = data.toByte()
      }
    }

  override fun overrideOpen(value: Boolean) {
    if (open != open) {
      data += if (open) 4 else -4
    }
  }
}
